dataflow/flex-templates/streaming_beam/e2e_test.py

```diff
@@ -0,0 +1,106 @@
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the 'License');
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an 'AS IS' BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+
+import json
+import time
+
+# `conftest` cannot be imported when running in `nox`, but we still
+# try to import it for the autocomplete when writing the tests.
+try:
+    from conftest import Utils
+except ModuleNotFoundError:
+    from typing import Any
+
+    Utils = Any
+import pytest
+
+NAME = "dataflow-flex-templates-streaming-beam"
+
+
+@pytest.fixture(scope="session")
+def bucket_name(utils: Utils) -> str:
+    yield from utils.storage_bucket(NAME)
+
+
+@pytest.fixture(scope="session")
+def pubsub_topic(utils: Utils) -> str:
+    yield from utils.pubsub_topic(NAME)
+
+
+@pytest.fixture(scope="session")
+def pubsub_subscription(utils: Utils, pubsub_topic: str) -> str:
+    yield from utils.pubsub_subscription(pubsub_topic, NAME)
+
+
+@pytest.fixture(scope="session")
+def bigquery_dataset(utils: Utils) -> str:
+    yield from utils.bigquery_dataset(NAME)
+
+
+@pytest.fixture(scope="session")
+def pubsub_publisher(utils: Utils, pubsub_topic: str) -> bool:
+    yield from utils.pubsub_publisher(
+        pubsub_topic,
+        new_msg=lambda i: json.dumps(
+            {
+                "url": "https://beam.apache.org/",
+                "review": "positive" if i % 2 == 0 else "negative",
+            }
+        ),
+    )
+
+
+@pytest.fixture(scope="session")
+def flex_template_image(utils: Utils) -> str:
+    yield from utils.container_image(NAME)
+
+
+@pytest.fixture(scope="session")
+def flex_template_path(utils: Utils, bucket_name: str, flex_template_image: str) -> str:
+    yield from utils.dataflow_flex_template_build(
+        bucket_name=bucket_name,
+        template_image=flex_template_image,
+        metadata_file="metadata.json",
+    )
+
+
+def test_flex_template_run(
+    utils: Utils,
+    bucket_name: str,
+    pubsub_publisher: str,
+    pubsub_subscription: str,
+    flex_template_path: str,
+    bigquery_dataset: str,
+) -> None:
+
+    bigquery_table = "output_table"
+    job_id = utils.dataflow_flex_template_run(
+        job_name=NAME,
+        template_path=flex_template_path,
+        bucket_name=bucket_name,
+        parameters={
+            "input_subscription": pubsub_subscription,
+            "output_table": f"{bigquery_dataset}.{bigquery_table}",
+        },
+    )
+
+    # Since this is a streaming job, it will never finish running.
+    # Wait for 10 minutes, and then cancel the job.
+    time.sleep(10 * 60)
```

Comment:
I don't love having sleeps in the test. Is there another way that we can wait until the job starts streaming and then cancel it, or is waiting a fixed amount of time the only way?

Reply:
I was waiting until some date came in and was processed to make sure the pipeline didn't explode and data is windowed correctly, etc. Unfortunately, this also takes into account the 2-5 minute startup of the worker VMs. I could make this shorter, to maybe 5-8 minutes. We could also poll the job until it's in a "running" state, and then sleep for 1-2 minutes just to allow some data to get processed.

Reply:
I like the second option to poll the job until it's in a running state and then sleep short amount of time as needed.

Reply:
I made a new wait function. It will poll every 30 seconds until the job is in a running state. It will poll for a maximum of 10 minutes which should be plenty of time for jobs to show up, but in case anything went wrong it won't loop forever. In most cases it should be running in around 2-3 minutes.